src/accounts.rs

```diff
@@ -31,84 +35,386 @@ pub struct ErrorCounters {
+        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
+        let p = format!("{}/{}", p, id);
+        let path = Path::new(&p);
+        let _ignored = remove_dir_all(path);
+        create_dir_all(path).expect("Create directory failed");
+        Arc::new(RwLock::new(AppendVec::<Account>::new(
+            &path.join(ACCOUNT_DATA_FILE),
+            true,
+            ACCOUNT_DATA_FILE_SIZE,
+            0,
+        )))
+    }
+
+    pub fn get_vote_accounts(&self, fork: Fork) -> Vec<Account> {
+        let mut accounts: Vec<Account> = vec![];
+        self.index_info
+            .vote_index
+            .read()
+            .unwrap()
+            .iter()
+            .for_each(|p| {
+                if let Some(forks) = self.index_info.index.read().unwrap().get(p) {
+                    for (v_fork, (id, index)) in forks.0.iter() {
+                        if fork == *v_fork {
+                            accounts.push(
+                                self.storage.read().unwrap()[*id]
+                                    .appendvec
+                                    .read()
+                                    .unwrap()
+                                    .get_account(*index)
+                                    .unwrap(),
+                            );
+                        }
+                        if fork > *v_fork {
+                            break;
+                        }
+                    }
+                }
+            });
+        accounts
+    }
+
+    pub fn hash_internal_state(&self, fork: Fork) -> Hash {
+        let mut ordered_accounts = BTreeMap::new();
+        let rindex = self.index_info.index.read().unwrap();
+        rindex.iter().for_each(|(p, forks)| {
+            for (v_fork, (id, index)) in forks.0.iter() {
+                if fork == *v_fork {
+                    let account = self.storage.read().unwrap()[*id]
+                        .appendvec
+                        .read()
+                        .unwrap()
+                        .get_account(*index)
+                        .unwrap();
+                    ordered_accounts.insert(*p, account);
+                }
+                if fork > *v_fork {
+                    break;
+                }
+            }
+        });
 
         hash(&serialize(&ordered_accounts).unwrap())
     }
 
-    fn load<U>(checkpoints: &[U], pubkey: &Pubkey) -> Option<Account>
-    where
-        U: Deref<Target = Self>,
-    {
-        for db in checkpoints {
-            if let Some(account) = db.accounts.get(pubkey) {
-                return Some(account.clone());
+    fn load(&self, fork: Fork, pubkey: &Pubkey) -> Option<Account> {
+        let index = self.index_info.index.read().unwrap();
+        if let Some(forks) = index.get(pubkey) {
+            // find most recent fork that is an ancestor of current_fork
+            for (v_fork, (id, offset)) in forks.0.iter() {
```

Comment:
how is the order enforced?  i am not sure if there is a guarantee that forks will be stored in order.  i think you need to scan the entire vector to find the highest fork that is equal to or less then the request